### ISSUE TITLE
Feature: ingress-nginx traffic management

### ISSUE THREAD

**Issue by moensch** [argoproj/argo-rollouts#401]
We intend on contributing ingress-nginx traffic management. Opening this issue to discuss the proposed design.

I propose we extend the [proposed spec](https://github.com/argoproj/argo-rollouts/blob/737d9cd537303ccf40c29056925da730a9763f07/docs/features/traffic-management/nginx.md#integration-with-argo-rollouts) with an additional `canaryIngress` field. this means the user has to create **two** `extensions/v1beta1.Ingress` resources, and pass both to the rollouts controller.
We should also rename `primaryIngress` to `stableIngress` to match `stableService`.

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Rollout
spec:
  ...
  strategy:
    canary:
      canaryService: canary-service # required
      stableService: stable-service  # required
      trafficRouting:
        nginx:
           annotationPrefix: example.nginx.com/ # optional
           stableIngress: stable-ingress  # required
```

The controller will then perform the following actions:

1. Validate the `Ingress` referenced by `stableIngress` is pointed at the `stableService ` in all host rules (`/rules[*]/http/paths[*]/backend/serviceName`)
2. Create a copy of the `stableIngress` (append `name` with `-canary`). In this copy, add the  `nginx.ingress.kubernetes.io/canary: "true"` annotation.
4. During reconcile, update the `nginx.ingress.kubernetes.io/canary-weight` annotation on the canary ingress to match the desired weight. Set to `0` once we reach `100`, similar as is done with the current Istio traffic management reconciler.

**Comment by dthomson25** [comment]
This is awesome to hear! Please let me know if you have any issues along the way!

For your proposal, what was your motivation to allow users to specify a reference to canary ingress? The value of specifying a reference is allowing the reference to be customized as needed while the Rollout object only makes the necessary changes to the reference object. In this case, however, the canary ingress needs to mirror the stable ingress in order for the canary functionality to work.  If a user specifies a canary ingress, that user needs to ensure that the canary spec matches the stable spec ingress. Even if the rollout finds them out of sync, the user is still responsible for reconciling those differences in the Canary and Stable ingress. We can automate that work for our users if the Rollout controller creates the canary Ingress. 

Do you have some other concerns with having the Canary ingress created by the Rollout controller?

**Comment by moensch** [comment]
@dthomson25 I probably should have posted a link to the original slack thread: https://argoproj.slack.com/archives/CKTN88XFH/p1580969954028600

Not mirroring the resources is in-line with the existing design approach around `Service` objects (where the user passes in two objects).

**Comment by jessesuen** [comment]
@dthomson25 and I just discussed this and we both agree that specifying a "canaryIngress" is probably unnecessary in the spec. 

I think probably confused the discussion in slack, when I had originally stated we did not care for the approach that flagger took where it creates these "shadow" copies of objects. However, this was in context of their approach to duplicating the Deployment and HPA objects, resulting in the originally deployed user-managed objects to effectively act as unused, dummy reference objects. In flagger, the "real" Deployment and HPA objects that actually serve production, are the ones with the `-primary` suffix which are created and owned by the Flagger Canary object. Our issue with this approach is the confusion around these extra objects.

However, in the context of nginx-ingress, the original Ingress object does not act as a dummy. The creation of the duplicate Ingress object is in fact a requirement from nginx-ingress for it to achieve canary. Furthermore, the second Ingress object itself is required to be identical to the primary ingress, which means there's not really a point in a user deploying this themselves, since they do not need to customize the second ingress. For this reason, we think it will make a better user experience if Argo Rollouts simply creates and owns the canary ingress itself.
